File: src/data_structures/linked_list/LinkedList.kt
package data_structures.linked_list

class LinkedList<T> {
    private var head: Node<T>? = null
    private var tail: Node<T>? = null
    var size: Int = 0

    fun isEmpty(): Boolean = size == 0

    fun addToFirst(value: T): LinkedList<T> {
        head = Node(value, head)
        if (tail == null) {
            tail = head
        }
        size++
        return this
    }

    fun addToEnd(value: T): LinkedList<T> {
        if (isEmpty()) {
            addToFirst(value)
            return this
        }
        tail?.next = Node(value)
        tail = tail?.next
        size++
        return this

    }

    fun nodeAt(index: Int): Node<T>? {
        var currentNode = head
        var currentIndex = 0

        while (currentNode != null && currentIndex < index) {
            currentNode = currentNode.next
            currentIndex++
        }
        return currentNode
    }

    fun insert(value: T, index: Int): LinkedList<T> {

        if (index >= size-1) return addToEnd(value)

        if (index <= 0) return addToFirst(value)

        val beforeNode = nodeAt(index - 1)
        val newNode = Node(value, beforeNode?.next)
        beforeNode!!.next = newNode
        size++
        return this

    }

    fun pop(): T? {
        if (!isEmpty()) size--

        val result = head?.value
        head = head?.next
        if (isEmpty()) tail = null

        return result
    }

    fun removeLast(): T? {
        val head = head ?: return null
        if (head.next == null) return pop()

        size--
        var pre = head
        var current = head

        var next = current.next
        while (next != null) {
            pre = current
            current = next
            next = current.next
        }
        pre.next = null
        tail = pre
        return current.value
    }

    fun removeAt(index: Int): T? {
        if (index >= size-1) return removeLast()
        if (index <= 0) return pop()

        val afterNode = nodeAt(index-1)
        val result = afterNode?.next?.value
         size--

        afterNode?.next = afterNode?.next?.next
        return result
    }


    override fun toString(): String {
        return if (isEmpty()) "Empty list" else head.toString()
    }

}

File: data_structures/linked_list/linked_list.py
from node import Node
import math

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    @property
    def is_empty(self):
        return self.size == 0

    def push(self, value):
        self.head = Node(value, self.head)
        if self.tail is None:
            self.tail = self.head
        self.size += 1
        return self

    def append(self, value):
        if self.is_empty:
            return self.push(value)
        self.tail.next = Node(value, self.tail.next)
        self.tail = self.tail.next
        self.size += 1
        return self

    def node_at(self, index):
        current_node = self.head
        current_index = 0
        while current_node is not None and current_index < index:
            current_node = current_node.next
            current_index += 1
        return current_node

    def insert(self, value, index):
        if index <= 0:
            return self.push(value)
        if index >= self.size:
            return self.append(value)

        after_node = self.node_at(index-1)
        node = Node(value, after_node.next)
        after_node.next = node
        self.size += 1
        return self

    def pop(self):
        if not self.is_empty:
            self.size -= 1
            result = self.head.value
            self.head = self.head.next
            if self.is_empty:
                self.tail = None
            return result

    def remove_last(self):
        if self.head is None:
            return None

        if self.head.next is None:
            return self.pop()

        self.size -= 1
        prev = self.head
        current = self.head
        next = current.next
        while next is not None:
            prev = current
            current = next
            next = next.next

        prev.next = None
        self.tail = prev
        return current.value

    def remove_last2(self):
        if self.head is None:
            return None

        if self.head.next is None:
            return self.pop()

        self.size -= 1
        last_node = self.node_at(self.size-1)
        last_node.next = None
        result = self.tail.value
        self.tail = last_node
        return result

    def remove_at(self, index):
        if index <= 0:
            return self.pop()
        if index >= self.size - 1:
            return self.remove_last()

        prev = self.head
        current = self.head
        next = current.next
        current_index = 0
        while next is not None and current_index < index:
            prev = current
            current = next
            next = current.next
            current_index += 1
        
        prev.next = next 
        self.size -=1
        return current.value

    def reverse(self):
        new_list = LinkedList()
        current = self.head
        while current is not None:
            new_list.push(current.value)
            current = current.next
        self.head = new_list.head
        self.tail = new_list.tail
        return self;

    def get_middle(self):
        return self.node_at(math.floor((self.size-1)/2)).value


    def to_string(self):
        if self.is_empty:
            return "Empty list"
        else:
            return self.head.to_string()
